File: 234-Palindrome-Linked-List.py
class Solution:
    def reverseList(self, head: ListNode) -> ListNode:
        prev, cur = None, head

        while cur:
            next = cur.next
            cur.next = prev
            prev = cur
            cur = next

        return prev

    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        if not head:
            return True

        slow, fast = head, head.next
        while fast and fast.next:
            slow, fast = slow.next, fast.next.next

        next_half, cur = self.reverseList(slow.next), head

        while next_half:
            if cur.val != next_half.val:
                return False
            cur, next_half = cur.next, next_half.next

        return True

File: 234-Palindrome-Linked-List.cpp
class Solution {
public:
    bool isPalindrome(ListNode* head) {
        ListNode* mid_node = findMid(head);
        ListNode* curr = head;

        mid_node = reverseList(mid_node);

        while (curr && mid_node) {
            if (curr->val != mid_node->val)
                return false;
            curr = curr->next;
            mid_node = mid_node->next;
        }

        return (true);
    }

private:
    ListNode* findMid(ListNode* head) {
        if (head == nullptr || head->next == nullptr)
            return head;

        ListNode* fast = head;
        ListNode* slow = head;

        while (fast->next && fast->next->next) {
            slow = slow->next;
            fast = fast->next->next;
        }

        return (slow->next);
    }

    ListNode* reverseList(ListNode* head) {
        ListNode* prev = nullptr;
        ListNode* curr = head;
        ListNode* next;

        while (curr) {
            next = curr->next;
            curr->next = prev;
            prev = curr;
            curr = next;
        }

        return (prev);
    }
};